### PR TITLE
Implement update rule status

### LINKED ISSUE #2558: router implement tracking list: custom message delivery between cloud and edge
**What would you like to be added/modified**:
- add custom message delivery machanism between cloud and edge

**Why is this needed**:
- Users need to deliver their custom messages between cloud and edge

**What have we done?**:
-  [x] support to deliver custom message from cloud to edge by mqttbroker. PR: #2501 
-  [x] support to deliver custom message from edge to cloud by mqttbroker. PR: #2501 

**What will we do?**:
-  [x] support to delivery custom message from cloud to edge by servicebus module.
-  [x] support to record messages' successful and failed count and failed reasons
-  [x] fix bug of delivering messages from edge to cloud when edgecore restarts.
-  [ ] support tls when deliverying messages from cloud to edge.

### PR DESCRIPTION
**Pull request by MesaCrush**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
For commit code.

**Which issue(s) this PR fixes**:
Fixes #2558 

**Special notes for your reviewer**:
Have a good day.

**Does this PR introduce a user-facing change?**:
NONE
```release-note

```

### CODE REVIEW

**Comment by MesaCrush** [comment]
@fisherxu Please merge this PR.

**Comment by fisherxu** [comment]
> @fisherxu Please merge this PR.

I'll take a look.

File: cloud/pkg/router/rule/statushandler.go [hunk 1]
```diff
@@ -18,81 +24,33 @@ type ErrorMsg struct {
...
-		}
-		timer.Reset(30 * time.Second)
-
-		go handleStatus(ruleStatus, errResults)
-
-		ruleStatus = make(map[string][2]int)
-		errResults = []ExecResult{}
-	}
-
+func do(stop chan bool) {
```
Comment (fisherxu) [inline cloud/pkg/router/rule/statushandler.go:35]: What's the meaning of do?

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 2]
```diff
@@ -42,19 +42,27 @@ import (
...
+	rulesv1 "github.com/kubeedge/kubeedge/cloud/pkg/apis/rules/v1"
+	crdClientset "github.com/kubeedge/kubeedge/cloud/pkg/client/clientset/versioned"
 	"github.com/kubeedge/kubeedge/cloud/pkg/common/client"
 	"github.com/kubeedge/kubeedge/cloud/pkg/common/modules"
+	"github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/controller"
 	"github.com/kubeedge/kubeedge/cloud/pkg/edgecontroller/config"
 	"github.com/kubeedge/kubeedge/cloud/pkg/edgecontroller/constants"
 	"github.com/kubeedge/kubeedge/cloud/pkg/edgecontroller/messagelayer"
 	"github.com/kubeedge/kubeedge/cloud/pkg/edgecontroller/types"
+	rule2 "github.com/kubeedge/kubeedge/cloud/pkg/router/rule"
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/controller/upstream.go:54]: rule2?
```suggestion
	"github.com/kubeedge/kubeedge/cloud/pkg/router/rule"
```

File: cloud/pkg/router/rule/statushandler.go (hunk 1 above)
Comment (fisherxu) [reply cloud/pkg/router/rule/statushandler.go:35]: Please use a meaningful name.

File: pkg/apis/componentconfig/cloudcore/v1alpha1/types.go [hunk 3]
```diff
@@ -306,6 +306,8 @@ type EdgeControllerLoad struct {
 	// DeletePodWorkers indicates the load of delete pod workers
 	// default 4
 	DeletePodWorkers int32 `json:"deletePodWorkers,omitempty"`
+
+	UpdateRuleStatusWorkers int32 `json:"UpdateRuleStatusWorkers,omitempty"`
```
Comment (fisherxu) [inline pkg/apis/componentconfig/cloudcore/v1alpha1/types.go:310]: Add comments

File: common/constants/default.go [hunk 4]
```diff
@@ -94,6 +94,7 @@ const (
 	DefaultQueryNodeWorkers                  = 4
 	DefaultUpdateNodeWorkers                 = 4
 	DefaultDeletePodWorkers                  = 4
+	DefaultUpdateRuleStatusWorkers           = 1
```
Comment (fisherxu) [inline common/constants/default.go:97]: Why is different from others?

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 5]
```diff
@@ -42,19 +42,27 @@ import (
...
 	common "github.com/kubeedge/kubeedge/common/constants"
 	edgeapi "github.com/kubeedge/kubeedge/common/types"
 )
 
 // SortedContainerStatuses define A type to help sort container statuses based on container names.
 type SortedContainerStatuses []v1.ContainerStatus
 
+type RuleStatus struct {
+	Status rulesv1.RuleStatus `json:"status"`
+}
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/controller/upstream.go:64]: Can we delete this struct? Use the `rulesv1.RuleStatus` directly.

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**